h/security/policy/_api_cookie.py

```diff
@@ -28,7 +30,18 @@ def handles(request: Request) -> bool:
         ) in COOKIE_AUTHENTICATABLE_API_REQUESTS
 
     def identity(self, request: Request) -> Identity | None:
+        try:
+            check_csrf_origin(request)
+        except BadCSRFOrigin:
+            return None
+
+        try:
+            check_csrf_token(request)
+        except BadCSRFToken:
+            return None
```

Comment:
Normally you can add CSRF-token protection to a view with `@view_config(require_csrf=True)` but we don't generally want API views to require CSRF protection: when you call these views from a script (authenticating with a bearer token), or using an auth client, etc you don't have to include a CSRF token. We only want the requests to need a CSRF token when they're using the API auth cookie to authenticate.

So I've implemented the CSRF checks here in `APICookiePolicy`, meaning any request authenticated with an API auth cookie will need a CSRF token regardless of endpoint.

Reply:
Returning `None` here makes the app act as if the user isn't authenticated so (depending on what the request tried to do) they'll get responses like 404s and "you don't have permission" errors.

Alternatively we could let the `BadCSRFOrigin` or `BadCSRFToken` exception fly here, catch it in an exception view, and sends some kind of error response (whatever we want) this is unique to these CSRF problems rather than to do with the particular endpoint that was called (and what that endpoint does if you try to call it when unauthenticated).

Reply:
It would make debugging API errors easier if the response was more specific. I frankly think it is worth accepting the slight security cost of this.

Reply:
I've changed this to send a **400 Bad Request** with `{"status": "failure", "reason": "CSRF validation failed: invalid or missing CSRF token."}` whenever a request to a cookie-authenticatable API endpoint has a valid API auth cookie but does not have a valid CSRF token: https://github.com/hypothesis/h/pull/8864/commits/4b6727c1ffb8dd995a6bff164ca9eb349d48cc4c

400 is the status that Pyramid uses by default for CSRF errors, see: <https://docs.pylonsproject.org/projects/pyramid/en/latest/_modules/pyramid/exceptions.html>.

Ideally I would add an end-to-end functional test for this error response but I was unable to get one working due to an apparent bug or unexpected behaviour in webtest. It wasn't worth the time to get to the bottom of it.